python/ray/data/tests/test_streaming_integration.py

```diff
@@ -0,0 +1,291 @@
+import pytest
+import time
+
+from typing import List, Any
+
+import ray
+from ray.data.context import DatasetContext
+from ray.data._internal.execution.interfaces import (
+    ExecutionOptions,
+    ExecutionResources,
+    RefBundle,
+)
+from ray.data._internal.execution.streaming_executor import (
+    StreamingExecutor,
+)
+from ray.data._internal.execution.operators.all_to_all_operator import AllToAllOperator
+from ray.data._internal.execution.operators.map_operator import MapOperator
+from ray.data._internal.execution.operators.input_data_buffer import InputDataBuffer
+from ray.data._internal.execution.util import make_ref_bundles
+from ray._private.test_utils import wait_for_condition
+from ray.data.tests.conftest import *  # noqa
+
+
+def make_transform(block_fn):
+    def map_fn(block_iter):
+        for block in block_iter:
+            yield block_fn(block)
+
+    return map_fn
+
+
+def ref_bundles_to_list(bundles: List[RefBundle]) -> List[List[Any]]:
+    output = []
+    for bundle in bundles:
+        for block, _ in bundle.blocks:
+            output.append(ray.get(block))
+    return output
+
+
+def test_pipelined_execution(ray_start_10_cpus_shared):
+    executor = StreamingExecutor(ExecutionOptions())
+    inputs = make_ref_bundles([[x] for x in range(20)])
+    o1 = InputDataBuffer(inputs)
+    o2 = MapOperator.create(make_transform(lambda block: [b * -1 for b in block]), o1)
+    o3 = MapOperator.create(make_transform(lambda block: [b * 2 for b in block]), o2)
+
+    def reverse_sort(inputs: List[RefBundle]):
+        reversed_list = inputs[::-1]
+        return reversed_list, {}
+
+    o4 = AllToAllOperator(reverse_sort, o3)
+    it = executor.execute(o4)
+    output = ref_bundles_to_list(it)
+    expected = [[x * -2] for x in range(20)][::-1]
+    assert output == expected, (output, expected)
+
+
+def test_e2e_option_propagation(ray_start_10_cpus_shared):
+    DatasetContext.get_current().new_execution_backend = True
+    DatasetContext.get_current().use_streaming_executor = True
+
+    def run():
+        ray.data.range(5, parallelism=5).map(
+            lambda x: x, compute=ray.data.ActorPoolStrategy(2, 2)
+        ).take_all()
+
+    DatasetContext.get_current().execution_options.resource_limits = (
+        ExecutionResources()
+    )
+    run()
+
+    try:
+        DatasetContext.get_current().execution_options.resource_limits.cpu = 1
+        with pytest.raises(ValueError):
+            run()
+    finally:
+        DatasetContext.get_current().execution_options.resource_limits.cpu = None
+
+
+def test_configure_spread_e2e(ray_start_10_cpus_shared):
+    from ray import remote_function
+
+    tasks = []
+
+    def _test_hook(fn, args, strategy):
+        if "map_task" in str(fn):
+            tasks.append(strategy)
+
+    remote_function._task_launch_hook = _test_hook
+    DatasetContext.get_current().use_streaming_executor = True
+    DatasetContext.get_current().execution_options.preserve_order = True
```

Comment:
Should these configs in this test and the others below get set back to their default at the end of the test, no matter whether it succeeds or fails? Otherwise these config changes will live for the lifetime of the process/rest of the tests in the module, which shouldn't be a big deal for `use_streaming_executor` but might be surprising for `execution_options.preserve_order`.

Reply:
Also, it might be a lot easier to have a `def enable_streaming_preserve_order()` fixture defined in this module or `conftest.py` that takes care of this, so it can be easily reused across tests without impacting the actual code of the tests. https://github.com/ray-project/ray/blob/56b79117af1c0a9c786c9e88c3f4d1311de84593/python/ray/data/tests/conftest.py#L320-L329

Reply:
Let me write a context manager for this.